pkg/instance/service.go

```diff
@@ -76,43 +80,267 @@ func (s service) SaveInstance(instance *model.DeploymentInstance) error {
 	}
 
 	deployment.Instances = append(deployment.Instances, instance)
-	err = s.validateDeployment(deployment)
+
+	_, err = s.validateNoCycles(deployment.Instances)
+	if err != nil {
+		return err
+	}
+
+	err = s.resolveParameters(deployment)
+	if err != nil {
+		return err
+	}
+
+	err = s.validateDeploymentParameters(deployment)
 	if err != nil {
 		return err
 	}
 
 	return s.instanceRepository.SaveInstance(instance)
 }
 
-func (s service) validateDeployment(deployment *model.Deployment) error {
-	stacks, err := s.collectStacks(deployment)
+func (s service) validateNoCycles(instances []*model.DeploymentInstance) (graph.Graph[string, *model.DeploymentInstance], error) {
+	g := graph.New(func(instance *model.DeploymentInstance) string {
+		return instance.StackName
+	}, graph.Directed(), graph.PreventCycles())
+
+	for _, instance := range instances {
+		err := g.AddVertex(instance)
```

Comment:
If `errors.Is(err, graph.ErrVertexAlreadyExists)` we should return something like `fmt.Errorf("failed adding instance for stack %q as one already exists.", instance.StackName)`.